Fix no longer supported git PR refs

viceice opened the pull request:
<!-- Please describe your pull request here. -->
This pr removes the usage of not and no longer supported bitbucket server git api.

fixes #287
fixes #302

<!--
To mark your pull request as work in progress please create it as a draft pull request
-->

### Your checklist for this pull request

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

<!--
Put an `x` into the [ ] to show you have filled the information below
Describe your pull request below
-->

wwuck commented:
Can we get a new release with this merged @Jetersen ?

viceice commented:
Works for me, file needs to be renamed to hpi. 💪 

[cloudbees-bitbucket-branch-source.zip](https://github.com/jenkinsci/bitbucket-branch-source-plugin/files/4347566/cloudbees-bitbucket-branch-source.zip)

Jetersen commented:
@viceice please do not upload files, use the once coming from the Jenkins CI

viceice commented:
ok, sorry and thanks for infomation

f-cramer commented:
I guess `com.cloudbees.jenkins.plugins.bitbucket.filesystem.BitbucketSCMFileSystem` should be changed with this PR as well, shouldn't it? In lines 179 and 181 are some more instances of ref/pull-requests ...

viceice commented:
ok, I'll look at it

Jetersen commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java:
```diff
@@ -176,9 +176,10 @@ public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head, @Ch
                 } else if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) {
-                    ref = "pull-requests/" + pr.getId() + "/merge";
+                    // No longer supported
+                    ref = null;
```
This makes me really sad, the one reason why Bitbucket does not make sense for any CI but their own.

viceice replied:
Is there any other possibillity to let jenkins do the merge?

viceice replied:
If i look at the code, jenkins is merging the pr into the target before build. So we only loose the lightweight checkout for merge?

Jetersen replied:
Jenkins does the merge but it does so on master by cloning the entire repo.
The reason for lightweight checkout is that it prevents the Jenkins master from cloning the entire repo.

viceice commented:
I've not found a possible api for checking bitbucket server version. Or there is no official api yet. 😞

Jetersen commented:
Perhaps located in the header response?

viceice commented:
nope

Jetersen commented:
Again Bitbucket server is a mess.

viceice commented:
ok, there is a possibillity to get the version [/rest/api/1.0/application-properties](https://docs.atlassian.com/bitbucket-server/rest/7.0.1/bitbucket-rest.html#idp85)

Jetersen commented:
the field could be transient on bitbucket server config

viceice commented:
Should I implement a bitbucket version based switch? The property is also found in anciend server version [5.16.0](https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html#idm8296895104)

Jetersen commented:
So far it is only v7 which has the issue newly introduced issue.

viceice commented:
ok, so I've checkd it again. The `refs/pull-requests/...` are no longer added since bitbucket server v7.

So `BitbucketSCMSource` can be changed as is, without version change.

The `build` function in `BitbucketSCMFileSystem.java` needs a bitbucket version check for `pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE` branch to support lightweight checkout on bitbucket server version below v7.

Right?

Jetersen commented:
Yup @viceice that would be much appreciated. My suggestion is to store the version once it is fetched on the bitbucket server config object.

viceice commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java:
```diff
@@ -176,8 +176,9 @@ public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head, @Ch
                 } else if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.HEAD) {
-                    ref = "pull-requests/" + pr.getId() + "/from";
+                    ref = "heads/" + pr.getBranchName();
                 } else if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) {
+                    // No longer supported since bitbucket server v7, falls back to heavycheckout
                     ref = "pull-requests/" + pr.getId() + "/merge";
                 }
```
@Jetersen What do you think? This supportes older bitbucket and falls back to heavy checkout on bitbucket v7 and later

Jetersen replied:
Yup seems fine with me, properly simpler fix 😅

jftsunami replied:
Older versions of BitBucket, while they had the pull-request refs, are not accurate. PRs get confused in our environment, and we're running 6.7.2.

viceice replied:
So the pr refs are not updated at 6.7.2 too?

jftsunami replied:
Not without performing a 'null' operation on the UI, such as editing and saving the PR without making any changes.

Jetersen replied:
or using the bitbucket webhook plugin that was actually doing this work for you.

https://marketplace.atlassian.com/apps/1215474/post-webhooks-for-bitbucket?hosting=server&tab=overview

jftsunami replied:
I'd recommend getting rid of the use of "refs/pull-requests" globally, as it was never a supported API by BitBucket, and only worked by accident or as a side effect, apparently.

Jetersen replied:
it has been working great for others so I have to disagree.

jftsunami replied:
Alright, but it did stop working before v7.  Not sure when.

jftsunami replied:
Would exposing a configuration option be a compromise?

viceice replied:
this condition only allows jenkins to do a lightweight checkout on master to fetch the jenkins file. When doing a full checkout on worker it would always fetch the correct commits with this pr.

So there is only a small window when jenkins fetches an out of date jenkinsfile. So this can only happen, if you push changes to an existing pr, which updates the jenkinsfile

jftsunami replied:
Oh, my mistake then.

jansohn replied:
I think this comment (https://jira.atlassian.com/browse/BSERV-12284?focusedCommentId=2389584&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-2389584) is a quite interesting read on the topic.
Taking this into account I would also get rid of that merge ref as it does not seem to come back and it can be outdated on earlier Bitbucket versions if people don't use special Bitbucket plugins that trigger the update of the refs, e.g. when using only the native web-hooks.
I don't know this source code well enough but we might want to add that REST API call (rest/api/1.0/projects/KEY/repos/slug/pull-requests/N/changes?limit=1) to make sure the from ref is properly created/updated if we still want to use it. This would allow users to stick with the native web-hooks solution without any additional plugins.
Thoughts?

viceice commented:
@Jetersen I have a idea to get this working on never bitbucket.

https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/master/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java#L1111-L1135

We can use the [diff](https://docs.atlassian.com/bitbucket-server/rest/7.0.1/bitbucket-rest.html#idp319) api to check for Jenkinsfile changes.

If `file.getRef().startsWith('refs/pull-request')` and jenkinsfile not changed, simply return target file.

maxime-michel commented:
Hey, thanks for the work being done here. Our CI workflow is broken after the upgrade to BB 7.0.1 and we're glad to see this conversation started. As we got told [here](https://jira.atlassian.com/browse/BSERV-12198?focusedCommentId=2373717&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2373717), `pull-requests/*/from` refs aren't updated automatically anymore.

Out of despair, we tried to run the release candidate posted above. It wasn't able to fetch the pull request at `refs/heads/<PR_BRANCH_NAME>`, so we had to rollback. 2.7.0 at least allowed us to manually click the Diff tab in Bitbucket, which would work around the issue. I just thought I'd share that before a release goes out.

viceice commented:
@maxime-michel Can you check if the `pull-requests/*/from` are updated after you use the diff api?

viceice commented:
Ok, tested it myself, even after viewing the pr diff the `pull-requests/*/merge` is missing

mkgl commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketGitSCMBuilder.java:
```diff
@@ -104,8 +104,9 @@ public BitbucketGitSCMBuilder(@NonNull BitbucketSCMSource scmSource, @NonNull SC
-                withRefSpec("+refs/pull-requests/" + pullId + "/from:refs/remotes/@{remote}/" + head.getName());
+                // https://community.atlassian.com/t5/Bitbucket-questions/Difference-of-refs-pull-requests-lt-ID-gt-merge-and-refs-pull/qaq-p/772142
+                String branchName = ((PullRequestSCMHead) head).getBranchName();
+                withRefSpec("+refs/heads/" + branchName + ":refs/remotes/@{remote}/" + branchName);
```
Not sure how this is supposed to handle PRs where source branch is in a fork?
Probably fine to do a branch lookup first, but maintain the `refs/pull-requests/x/from` as a fallback?

viceice replied:
shouldn't the `@{remote}` be the fork? 🤔

mkgl replied:
`@{remote}` is only used in the "receiving" part of the refspec, after the colon, which just indicates how it's gonna store that ref "locally"; that's irrelevant with where it's supposed to find the source branch. Also if we were on the fork already, we wouldn't have to use those pull-requests refs in the first place. :)

Besides navigating through, as `#withRefSpec` indicates, `@{remote}` is supposed to be processed for substitution by `#remote()` before use, which is only ever gonna return a fixed remote for the current scm source, as far as I can tell.

viceice replied:
I havn't tested with forks, because we don't use them. Maybe i can try it.

JeroennC replied:
We've just tested plugin version `2.7.1-rc629.f242e60a14b` with our setup which uses forks, and indeed it fails at this step as the branches can not be found on the origin

chq81 commented:
@viceice Hi there. Any progess or ideas when this can be fixed. My company updated to Bitbucket v7 without giving prior notice and now all my build pipeline/workflow is broken due to the missing PR functionality.

Jetersen commented:
Again I point the blame at Atlassian for breaking things.

I wonder how their bitbucket server plugin handles these refs.

chq81 commented:
Thanks @Jetersen. I didn't mean to imply that someone needs to fix this plugin, and I totally understand that when Atlassian changes the way of how they handle PRs, it is their fault. It just seemed that you guys are working on a solution, therefore I thought I'll ask.

Jetersen commented:
No worries @chq81 :)

Checked out the Atlassian Bitbucket server plguin and they do not even offer support for these features 🤷‍♂ I still do not understand their need to create a separate plugin.

maxime-michel commented:
My understanding at this point is that you'd use the Atlassian plugin to connect Jenkins and Bitbucket together (with the automatically created webhook), but that for PRs you would use: https://github.com/jenkinsci/bitbucket-push-and-pull-request-plugin/

viceice commented:
> @viceice Hi there. Any progess or ideas when this can be fixed. My company updated to Bitbucket v7 without giving prior notice and now all my build pipeline/workflow is broken due to the missing PR functionality.

I'm using this pr build in production, which works so far.

chq81 commented:
@viceice Thanks for the hint. I installed it and it works so far. Would still be nice so see it as an official release, but I guess there are still other problems to be solved before this can be merged... In addition to Atlassian either rebuilding their original functionality or adding PRs into their own plugin.

jetersen commented:
Could someone test #302 with Can Merge enabled?

maxime-michel commented:
Yes, let me do that. I'll report later today.

maxime-michel commented:
Works for our scenario! Without the change:

1. PR is created from branch of fork against canonical's master
2. canonical's Jenkins job scan is triggered, but Jenkinsfile not found for PR
3. temporary workaround of clicking the 'Diff' tab for the PR in Bitbucket
4. manual repository scan, PR created

With this change:

1. 'Call can merge' enabled
2. PR is created from branch of fork against canonical's master
3. canonical's Jenkins job scan is triggered, Jenkinsfile found for PR
4. PR is built successfully, despite the following warnings:

```    ERROR: Could not do lightweight checkout, falling back to heavyweight
    java.io.FileNotFoundException: URL: /rest/api/1.0/projects/modules/repos/log-tools/browse/Jenkinsfile?at=pull-requests%2F33%2Fmerge&start=0&limit=500
            at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getRequest(BitbucketServerAPIClient.java:831)
            at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getFileContent(BitbucketServerAPIClient.java:1123)
            at com.cloudbees.jenkins.plugins.bitbucket.filesystem.BitbucketSCMFile.content(BitbucketSCMFile.java:98)
            at jenkins.scm.api.SCMFile.contentAsString(SCMFile.java:335)
            at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:107)
            at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:309)
            at hudson.model.ResourceController.execute(ResourceController.java:97)
            at hudson.model.Executor.run(Executor.java:428)
```

``` > git checkout -f a79164d0a991257c7080fa7f7b722a127d696a68 # timeout=10
 > git remote # timeout=10
 > git config --get remote.origin.url # timeout=10
…
 > git merge 7ab7a6482e437e87d7416f180134cbccd7a41c7e # timeout=10
 > git rev-parse HEAD^{commit} # timeout=10
Merge succeeded, producing a79164d0a991257c7080fa7f7b722a127d696a68
Checking out Revision a79164d0a991257c7080fa7f7b722a127d696a68 (PR-33)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f a79164d0a991257c7080fa7f7b722a127d696a68 # timeout=10
Commit message: "test-pr-2"
…
fatal: bad object a79164d0a991257c7080fa7f7b722a127d696a68
```

Rooomchik commented:
Hello Everyone. I'm a newbie in such a process so sorry for something is wrong.

Our company has upgraded the Bitbucket to 7 version and we are facing the same issue.

Can you please advise if there are HL timelines fix that will be released with a newer version of the plugin?

mlasevich commented:
For what its worth, I back-ported these changes against [2.5.0](https://github.com/mlasevich/bitbucket-branch-source-plugin/tree/feature/2.5.0-with-bb7) (some of us are stuck on older version of plugin thanks to CloudBees) and getting the same results as above - the API error is still there but it seems to work.

asafp-nice commented:
Hi, we've also tested this PR and it indeed appears to fix the issue for us. We're using Bitbucket server 7.0.3.

Any plans on merging this to Master?

jetersen commented:
As many have mentioned there is still the issue of lightweight checkout with the `FileNotFoundException` which would be great if it could get fixed.

As I do not use Bitbucket I have no way to figure out the solution.

Would prefer if the community could find a solution. I'll be happy to review and merge any PR that solves the issue.

Please see #288

jetersen commented:
https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/294#discussion_r420672772
With this comment we definitely need a better approach.

viceice commented:
I think heavyweight checkout (bitbucket > v7) is required for testing merged pr's when Jenkinsfile changed. Otherwise we could fallback to the target Jenkinsfile.

The merge commit will never come back to bitbucket, so we need a possibility to get files from source branch, especially if it is from a fork.

jetersen commented:
I think we should fallback to target Jenkinsfile, unless perhaps otherwise configured to use source if configured to trust collaborator?

viceice commented:
@jetersen Any idea why i get this error locally?
![image](https://user-images.githubusercontent.com/1798109/81172431-86f48d00-8f9e-11ea-970a-0effd1d04704.png)

jetersen commented:
@viceice you need to use java 8 to build.

JeroennC commented:
@jetersen - We also face issues that new PRs are not triggered due to `'Jenkinsfile' not found` on the multibranch pipeline scan. Wondering if you can assess with your domain knowledge whether this PR also applies to this issue.

I will wait for the incremental build to be available and test this, create a new issue if it persists.

viceice commented:
I've reverted most changes now and added the suggested changes api call. This will hopefully bring back the `from` ref, but does not work for `merge` ref

jetersen commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java:
```diff
@@ -337,11 +338,16 @@ public String getRepositoryUri(@NonNull BitbucketRepositoryType type,
+        if (endpointConfig instanceof BitbucketServerEndpoint) {
             // This is required for Bitbucket Server to update the refs/pull-requests/* references
             // See https://community.atlassian.com/t5/Bitbucket-questions/Change-pull-request-refs-after-Commit-instead-of-after-Approval/qaq-p/194702#M6829
             for (BitbucketServerPullRequest pullRequest : pullRequests) {
```
Would prefer we avoid looping over these pull requests unless either `isCallCanMerge` or `isCallChanges` is enabled

jetersen commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java:
```diff
@@ -337,11 +338,16 @@ public String getRepositoryUri(@NonNull BitbucketRepositoryType type,
             // See https://community.atlassian.com/t5/Bitbucket-questions/Change-pull-request-refs-after-Commit-instead-of-after-Approval/qaq-p/194702#M6829
             for (BitbucketServerPullRequest pullRequest : pullRequests) {
-                pullRequest.setCanMerge(getPullRequestCanMergeById(Integer.parseInt(pullRequest.getId())));
+                if (((BitbucketServerEndpoint) endpointConfig).isCallCanMerge()) {
+                    pullRequest.setCanMerge(getPullRequestCanMergeById(Integer.parseInt(pullRequest.getId())));
+                }
+                if (((BitbucketServerEndpoint) endpointConfig).isCallChanges()) {
+                    callPullRequestChangesById(Integer.parseInt(pullRequest.getId()));
+                }
```
Better to only cast it once.
Please cast the endpoint once outside the for loop
```java
                BitbucketServerEndpoint endpoint = (BitbucketServerEndpoint) endpointConfig;
```
```suggestion
                if (endpoint.isCallCanMerge()) {
                    pullRequest.setCanMerge(getPullRequestCanMergeById(Integer.parseInt(pullRequest.getId())));
                }
                if (endpoint.isCallChanges()) {
                    callPullRequestChangesById(Integer.parseInt(pullRequest.getId()));
                }
```

viceice commented:
Testing the build from [here](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fbitbucket-branch-source-plugin/detail/PR-294/15/artifacts) in our jenkins and it seem to work (heavy checkout on merge)

Can somebody test this too? I'll do the requested suggestions now.

viceice commented:
@jetersen Shouln't i remove the integer parse? As we convert them to string anyways after replacing in `UrlTemplate`

jetersen commented:
For the new addition, yes.
For the canMerge an extension plugin is using them as int so that would be a breaking change.

jetersen commented:
Nevermind, since both are private go for it. Feel free to remove the parsing of the ids.

```java
    private void callPullRequestChangesById(@NonNull Integer id) throws IOException {
        String url = UriTemplate
                .fromTemplate(API_PULL_REQUEST_CHANGES_PATH)
                .set("owner", getUserCentricOwner())
                .set("repo", repositoryName)
                .set("id", id).set("limit", 1)
                .expand();
        getRequest(url);
    }

    private boolean getPullRequestCanMergeById(@NonNull Integer id) throws IOException {
        String url = UriTemplate
                .fromTemplate(API_PULL_REQUEST_MERGE_PATH)
                .set("owner", getUserCentricOwner())
                .set("repo", repositoryName)
                .set("id", id)
                .expand();
        String response = getRequest(url);
        try {
            return JsonParser.toJava(response, BitbucketServerPullRequestCanMerge.class).isCanMerge();
        } catch (IOException e) {
            throw new IOException("I/O error when accessing URL: " + url, e);
        }
    }
```

viceice commented:
ok, done

JeroennC commented:
Tested this on our setup - with forks - and it works 👍 
Falls back to heavy weight checkout to get the Jenkinsfile correctly, also now during the multibranch pipeline scan it immediately finds the Jenkinsfile and starts a build as expected.

jetersen commented:
@viceice mind updating the user docs to mention bitbucket server 7 and suggest enabling the call changes api?

viceice commented:
@jetersen Can you please hint me where to put the `callChanges` stuff

jetersen commented:
@viceice feel free to update the remove the current important block.
And replace it with a mention on bitbucket server 7 and link to a new section you feel free to add the section at the top.

mattiassluis commented:
@jetersen I contacted @viceice but he is squeezed for time. I create the doc changes here: https://github.com/mattiassluis/bitbucket-branch-source-plugin/commit/26fbc94a411c0527f3acd4d81f3dd1b62223797c and I can either create a new PR that is the same as this one or you can maybe apply those changes to this one.

jetersen commented:
Let me cherry-pick it in :)

viceice commented:
@mattiassluis thanks

mattiassluis commented:
> Let me cherry-pick it in :)

Thank you! What's left now to get this in an official release?

viceice commented on docs/USER_GUIDE.adoc:
```diff
@@ -23,6 +22,14 @@ as a multi-branch project source in two different ways:
+[id=bitbucket-server-7]
+== BitBucket Server 7 compatibility
+
+_BitBucket Server 7.x_ is supported but does no longer support lightweight checkout for pull requests.
```
```suggestion
_BitBucket Server 7.x_ is supported but does no longer support lightweight checkout for pull requests when merge strategy is used for build.
```

Lightweight checkout works if we don't use merge build strategy.

mattiassluis replied:
Ah I said that correct on the first part but it is missing here indeed. Good catch!

mjholtkamp commented:
This looks great! Good work everyone! Is there something that needs to be done before a release? I'm looking forward to having this fixed :)

serg-bs commented:
1. I build plugin 2.8.1-SNAPSHOT (private-98e3c549-) 
Atlassian Bitbucket v7.3.0-m1
Jenkins ver. 2.222.3
1. Install plugin manually  in jenkins. 

1. Create multibranch job with bitbucket source and Behaviours - Discover pull requests from origin "Merging the pull request with the current target branch revision"

And still have `ERROR: Could not do lightweight checkout, falling back to heavyweight
java.io.FileNotFoundException: URL: /rest/api/1.0/projects/DI/repos/goodwork/browse/Jenkinsfile?at=pull-requests%2F1%2Fmerge&start=0&limit=500` 
look like [#287 ]reproduce. What should be wrong?

viceice commented:
@serg-bs As previously disscussed, you can ignore that error and it should do heavy checkout and work with bitbucket server >=7.0

mattiassluis commented:
@serg-bs as you can read in the related ticket this is a problem that cannot be solved. Starting Bitbucket 7 you can no longer perform the lightweight checkout on the merge because it simply no longer exists in Bitbucket. The jobs now luckily proceed with the heavyweight and still work.
The updated docs of this plugin (in this PR) also state this: https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/294/files#diff-8528ab6ef1a37be2a6df68940c0ad37bR7

MichaelKorn commented:
I installed the plugin manually from https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fbitbucket-branch-source-plugin/detail/PR-294/20/artifacts

Now, the first run of a PR looks always good with Bitbucket v7.2.3. But no further runs (if the Branch is updated) are possible, due to the following error:
```
Branch event
ERROR: Could not do lightweight checkout, falling back to heavyweight
java.io.FileNotFoundException: URL: /rest/api/1.0/projects/DBP/repos/reclis/browse/Jenkinsfile?at=pull-requests%2F8%2Fmerge&start=0&limit=500
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getRequest(BitbucketServerAPIClient.java:855)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getFileContent(BitbucketServerAPIClient.java:1150)
	at com.cloudbees.jenkins.plugins.bitbucket.filesystem.BitbucketSCMFile.content(BitbucketSCMFile.java:98)
	at jenkins.scm.api.SCMFile.contentAsString(SCMFile.java:335)
	at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:107)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:309)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:428)
Checking out git https://.....com/scm/DBP/reclis.git https://.....com/scm/DBP/reclis.git into /var/jenkins_home/workspace/al-Business-Platform_reclis_PR-8@script to read Jenkinsfile
using credential d93f1c48-792f-4a65-bba4-4f75ebcb9bea
using credential d93f1c48-792f-4a65-bba4-4f75ebcb9bea
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from 2 remote Git repositories
 > git config remote.origin.url https://.....com/scm/DBP/reclis.git # timeout=10
Fetching upstream changes from https://.....com/scm/DBP/reclis.git
 > git --version # timeout=10
using GIT_ASKPASS to set credentials CD/CI User
 > git fetch --tags --progress -- https://.....com/scm/DBP/reclis.git +refs/pull-requests/8/from:refs/remotes/origin/PR-8 # timeout=10
 > git config remote.upstream.url https://.....com/scm/DBP/reclis.git # timeout=10
Fetching upstream changes from https://......com/scm/DBP/reclis.git
using GIT_ASKPASS to set credentials CD/CI User
 > git fetch --tags --progress -- https://......com/scm/DBP/reclis.git +refs/heads/stage:refs/remotes/upstream/stage # timeout=10
Merging remotes/upstream/stage commit e5c90868f06713caa25fa65e12bb834f80da4a73 into PR head commit d1b96a1867121d6f0e9f28fb1f017738a17b6897
 > git config core.sparsecheckout # timeout=10
 > git checkout -f d1b96a1867121d6f0e9f28fb1f017738a17b6897 # timeout=10
ERROR: Checkout failed
hudson.plugins.git.GitException: Command "git checkout -f d1b96a1867121d6f0e9f28fb1f017738a17b6897" returned status code 128:
stdout: 
stderr: fatal: reference is not a tree: d1b96a1867121d6f0e9f28fb1f017738a17b6897

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2430)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$1100(CliGitAPIImpl.java:81)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:2743)
Caused: hudson.plugins.git.GitException: Could not checkout d1b96a1867121d6f0e9f28fb1f017738a17b6897
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$9.execute(CliGitAPIImpl.java:2767)
	at jenkins.plugins.git.MergeWithGitSCMExtension.checkout(MergeWithGitSCMExtension.java:144)
	at jenkins.plugins.git.MergeWithGitSCMExtension.decorateRevisionToBuild(MergeWithGitSCMExtension.java:110)
	at hudson.plugins.git.GitSCM.determineRevisionToBuild(GitSCM.java:1063)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1168)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:125)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:155)
	at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:142)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:309)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:428)
[Bitbucket] Notifying pull request build result
[Bitbucket] Build result notified
ERROR: Maximum checkout retry attempts reached, aborting
Finished: FAILURE
```
Ich suppose the cached git repository should be reset. As workaround a delete jenkins_home/workspace/[PR-FOLDER] on the jenkins master and one additional run is possible. Before the next run I have to delete it again.

viceice commented:
@MichaelKorn did you checked the `Call Changes api` checkbox ?

MichaelKorn commented:
Unfortunately I missed the note regarding the checkbox, sorry!
Now it's working. THX @viceice

wejrox commented:
Any news on this PR being merged? Or when a new release will be if that's what we're waiting for?

jetersen commented:
If someone else wants it released I'd suggest asking for maintainership.
[Please read about adopting a plugin](https://www.jenkins.io/doc/developer/plugin-governance/adopt-a-plugin/)
Also see https://github.com/jenkinsci/bitbucket-branch-source-plugin/issues/288

I for one cannot comfortably release this PR without working lightweight checkout for existing users.

joppa27 commented:
@jetersen I just want to take time to thank you & the rest of your team for all the work that you have done on this plugin.  It has made my life as a Bitbucket Server customer so much easier, and given me the flexibility to adopt different CI workflows.  I can only imagine how frustrating it is for you all that this plugin has essentially been broken by changes you had no control over.

I don't fully understand why Atlassian chose to make their own plugin instead of refactoring this one. The new plugin they made currently lacks a lot of the functionality I've come to depend on that you and your team made. I don't think Atlassian fully thought this through, and that decision is going to result in a lot of work for their customers. 

You & your team should not be responsible for Atlassians's decisions & I fully support you putting this plugin up for adoption. In my opinion that's the only logical choice.  If there's a place where I can donate some money or buy your and your team some beers please let me know.

Again Thank you all for all the work you have done and continue to do!! :)

joppa27 commented:
A well deserved Thank you goes out to Cloudbees too!

As much as I need/want the fix to be merged, stepping up to be a maintainer when the direction Atlassian wants to go in isn't clear doesn't make sense. The least I can do is sponsor you, and hope a resolution happens soon whatever that resolution happens to be.

faizan2131993 commented:
Thanks, Cloudbees, and everyone who has worked on the above issue to be gets fixed.

I see the changes have been approved. Do you have any plans to release a new version with this fix to the Jenkins market? This fix is really great and it may help many of us with this fix.

I appreciate the kind works here. Thank you.

jglick commented:
Have been running https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/294/commits/98e3c549af772f1e40959711bc8fac5ed9b50561 on some play projects in BB Server 7. Works if you check the new **Call changes API** checkbox AFAICT.

kshultzCB commented:
Is anyone aware of a reason why the setting this checkbox enables shouldn't be turned on all the time?

I have not done a full compatibility matrix of tests by any stretch (different combinations of settings, Cloud versus Server, forks, various branches, etc), but I did try this PR against Bitbucket 5 running in a docker container locally. The checkbox had no apparent ill effects on this basic test:

```
Branch indexing
Obtained Jenkinsfile from 35db29eccaafe28acac65a19b0ffe5d5c23cfe1d+62b7ff7d2c619f5d7abf90c6317b610668d07084
Running in Durability level: MAX_SURVIVABILITY
[Pipeline] Start of Pipeline
[Pipeline] node
Running on Jenkins in /Users/kshultz/GitHub/bitbucket-branch-source-plugin/work/workspace/checked5_PR-2
[Pipeline] {
[Pipeline] stage
[Pipeline] { (A Stage)
[Pipeline] echo
Basic Jenkinsfile
[Pipeline] sh
+ whoami
kshultz
[Pipeline] echo
Goodbye
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
[Bitbucket] Notifying pull request build result
[Bitbucket] Build result notified
Finished: SUCCESS
```
